Custom u32 index fails to compile despite implementing From<usize> and Into<usize>

Trying to build the following code:

```rs
use typed_index_collections::{TiVec, ti_vec};

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
struct CustomIndex(u32);

impl From<usize> for CustomIndex {
    fn from(value: usize) -> Self {
        Self(u32::try_from(value).expect("values above 2^32 - 1 aren't supported"))
    }
}

impl Into<usize> for CustomIndex {
    fn into(self) -> usize {
        self.0 as usize
    }
}

fn main() {
    let v: TiVec<CustomIndex, i32> = ti_vec![1, 2, 3];
    let idx = CustomIndex(2);
    assert_eq!(v[idx], 3);
}
```

fails to compile with this error:

```
error[E0277]: the trait bound `usize: From<CustomIndex>` is not satisfied
   --> src/main.rs:21:16
    |
 21 |     assert_eq!(v[idx], 3);
    |                ^^^^^^ the trait `From<CustomIndex>` is not implemented for `usize`
    |
help: the following other types implement trait `From<T>`
   --> /home/user/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/convert/num.rs:91:1
    |
 91 | impl_from!(bool => usize);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^ `usize` implements `From<bool>`
...
104 | impl_from!(u8 => usize, #[stable(feature = "lossless_int_conv", since = "1.5.0")]);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `usize` implements `From<u8>`
...
140 | impl_from!(u16 => usize, #[stable(feature = "lossless_iusize_conv", since = "1.26.0")]);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `usize` implements `From<u16>`
    |
   ::: /home/user/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/alignment.rs:206:1
    |
206 | impl const From<Alignment> for usize {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `usize` implements `From<std::ptr::Alignment>`
    = note: required for `CustomIndex` to implement `TiSliceIndex<CustomIndex, i32>`
    = note: required for `TiVec<CustomIndex, i32>` to implement `Index<CustomIndex>`
    = note: this error originates in the macro `impl_from` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
```

---

Changing the `Into<usize> for CustomIndex` impl to a `From<CustomIndex> for usize` impl makes it work, though it isn't obvious that breaking the orphan rule is allowed in this case. Clippy can tell you this, but only if your code compiles.

--- a/src/slice/slice_index.rs
+++ b/src/slice/slice_index.rs
@@ -77,40 +77,40 @@ pub trait TiSliceIndex<K, V>: private::Sealed<K> {
 
 impl<K, V> TiSliceIndex<K, V> for K
 where
-    usize: From<K>,
+    K: Into<usize>,
 {
     type Output = V;
 
     #[inline]
     fn get(self, slice: &TiSlice<K, V>) -> Option<&Self::Output> {
-        slice.raw.get(usize::from(self))
+        slice.raw.get(Into::<usize>::into(self))
     }
 
     #[inline]
     fn get_mut(self, slice: &mut TiSlice<K, V>) -> Option<&mut Self::Output> {
-        slice.raw.get_mut(usize::from(self))
+        slice.raw.get_mut(Into::<usize>::into(self))
     }
 
     #[inline]
     unsafe fn get_unchecked(self, slice: &TiSlice<K, V>) -> &Self::Output {
         // SAFETY: Guaranteed by the caller.
-        unsafe { slice.raw.get_unchecked(usize::from(self)) }
+        unsafe { slice.raw.get_unchecked(Into::<usize>::into(self)) }
     }
 
     #[inline]
     unsafe fn get_unchecked_mut(self, slice: &mut TiSlice<K, V>) -> &mut Self::Output {
         // SAFETY: Guaranteed by the caller.
-        unsafe { slice.raw.get_unchecked_mut(usize::from(self)) }
+        unsafe { slice.raw.get_unchecked_mut(Into::<usize>::into(self)) }
     }
 
     #[inline]
     fn index(self, slice: &TiSlice<K, V>) -> &Self::Output {
-        &slice.raw[usize::from(self)]
+        &slice.raw[Into::<usize>::into(self)]
     }
 
     #[inline]
     fn index_mut(self, slice: &mut TiSlice<K, V>) -> &mut Self::Output {
-        &mut slice.raw[usize::from(self)]
+        &mut slice.raw[Into::<usize>::into(self)]
     }
 }
 

